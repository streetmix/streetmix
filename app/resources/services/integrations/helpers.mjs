import models from '../../../db/models/index.js'

const { User, UserConnections, sequelize } = models

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
export async function findUser (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
export function addOrUpdateByProviderName (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
export async function syncAccountStatus (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

/**
 * When a user has successfully connected their Streetmix account with a third-
 * party identity provider, record that connected identity with the user's
 * Streetmix data in the UserConnection table.
 *
 * @todo Remove first attempt at storing identity information as a JSON blob
 *        on a User model
 * @param {Object} account - User's Streetmix account information
 * @param {Object} profile - User's connected identity information
 * @returns {Promise}
 */
export function addUserConnection (account, profile) {
  const identities = account.identities
  const identity = {
    provider: profile.provider,
    user_id: profile.id,
    access_token: profile.access_token || undefined, // Coil
    refresh_token: profile.refresh_token || undefined // Coil
  }

  // if provider exists, should update that item
  // if not, should add to the object list
  // TODO: Deprecated
  addOrUpdateByProviderName(identities, identity)

  return sequelize.transaction(async (t) => {
    // This is a previous attempt, where we saved connected identity
    // information to a JSON field in the User model. A newer method (using
    // the UserConnection table) is used below. Both are currently used to
    // test concurrent data validity. In time, the User.update() method should
    // be removed.
    await User.update(
      {
        identities
      },
      { where: { auth0Id: account.auth0Id }, returning: true, transaction: t }
    )

    await UserConnections.findOrCreate({
      where: {
        user_id: account.id,
        provider: profile.provider
      },
      transaction: t
    })
    return await UserConnections.update(
      {
        provider_user_id: profile.id,
        metadata: profile
      },
      {
        where: {
          user_id: account.id,
          provider: profile.provider
        },
        returning: true,
        transaction: t
      }
    )
  })
}
