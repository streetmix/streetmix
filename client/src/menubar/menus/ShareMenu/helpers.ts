import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
