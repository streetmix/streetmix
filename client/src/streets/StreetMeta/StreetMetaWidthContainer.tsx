import React, { useState } from 'react'
import { useIntl } from 'react-intl'
import { useSelector, useDispatch } from '../../store/hooks'
import { updateStreetWidthAction as updateStreetWidth } from '../../store/actions/street'
import {
  SETTINGS_UNITS_IMPERIAL,
  SETTINGS_UNITS_METRIC
} from '../../users/constants'
import { updateUnits } from '../../users/localization'
import { processWidthInput, prettifyWidth } from '../../util/width_units'
import {
  MIN_CUSTOM_STREET_WIDTH,
  MAX_CUSTOM_STREET_WIDTH,
  MIN_CUSTOM_STREET_WIDTH_IMPERIAL,
  MAX_CUSTOM_STREET_WIDTH_IMPERIAL,
  STREET_WIDTH_CUSTOM,
  STREET_WIDTH_SWITCH_TO_METRIC,
  STREET_WIDTH_SWITCH_TO_IMPERIAL
} from '../constants'
import { normalizeStreetWidth } from '../width'
import StreetMetaWidthLabel from './StreetMetaWidthLabel'
import StreetMetaWidthMenu from './StreetMetaWidthMenu'

function StreetMetaWidthContainer (): React.ReactElement {
  const street = useSelector((state) => state.street)
  const editable = useSelector(
    (state) => !state.app.readOnly && state.flags.EDIT_STREET_WIDTH.value
  )
  const locale = useSelector((state) => state.locale.locale)
  const dispatch = useDispatch()
  const [isEditing, setEditing] = useState(false)
  const intl = useIntl()

  /**
   * When the street width label is clicked, only allow editing if
   * street width is not read-only
   */
  const handleClickLabel = (event: React.MouseEvent): void => {
    if (editable) {
      setEditing(true)
    }
  }

  /**
   * Handles changes to the <select> dropdown rendered in
   * <StreetMetaWidthMenu />
   */
  const handleChangeMenuSelection = (value: string): void => {
    setEditing(false)

    const { units, width, occupiedWidth } = street
    const selection = Number.parseFloat(value)

    switch (selection) {
      case STREET_WIDTH_SWITCH_TO_METRIC:
        updateUnits(SETTINGS_UNITS_METRIC)
        break
      case STREET_WIDTH_SWITCH_TO_IMPERIAL:
        updateUnits(SETTINGS_UNITS_IMPERIAL)
        break
      // Prompt for new street width
      case STREET_WIDTH_CUSTOM: {
        const minValue =
          units === SETTINGS_UNITS_IMPERIAL
            ? MIN_CUSTOM_STREET_WIDTH_IMPERIAL
            : MIN_CUSTOM_STREET_WIDTH
        const maxValue =
          units === SETTINGS_UNITS_IMPERIAL
            ? MAX_CUSTOM_STREET_WIDTH_IMPERIAL
            : MAX_CUSTOM_STREET_WIDTH
        const promptValue = normalizeStreetWidth(occupiedWidth, units)
        const promptString = intl.formatMessage(
          {
            id: 'prompt.new-width',
            defaultMessage: 'New street width (from {minWidth} to {maxWidth}):'
          },
          {
            minWidth: prettifyWidth(minValue, units, locale),
            maxWidth: prettifyWidth(maxValue, units, locale)
          }
        )
        const inputWidth = window.prompt(
          promptString,
          prettifyWidth(promptValue, units, locale)
        )

        if (inputWidth !== null) {
          const newWidth = normalizeStreetWidth(
            processWidthInput(inputWidth, units),
            units
          )
          void dispatch(updateStreetWidth(newWidth))
        }

        break
      }
      // Do nothing if the selection is the original width
      case width:
        break
      // Change width to the desired selection
      default:
        if (selection) {
          void dispatch(updateStreetWidth(selection))
        }
        break
    }
  }

  return isEditing
    ? (
      <div className="street-meta-width">
        <StreetMetaWidthMenu
          street={street}
          onChange={handleChangeMenuSelection}
        />
      </div>
      )
    : (
      <StreetMetaWidthLabel
        street={street}
        editable={editable}
        onClick={handleClickLabel}
      />
      )
}

export default StreetMetaWidthContainer
