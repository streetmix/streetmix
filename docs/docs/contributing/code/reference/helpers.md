---
sidebar_position: 2
---

# Helpers and utilities

:::caution Under construction

This page is a work in progress.

:::

We use the following utilities to solve common programming problems.

## Pseudo-random number generator

If you need a one-time use random number (which will never be stored), there's nothing wrong with JavaScript's built-in `Math.random()`. However, there are times where we need a **seed** for a pseudo-random number generator (PRNG). Seeds generate a random-_looking_ sequence of numbers, but they have the benefit of remaining consistent across renders. For example, this is useful on our sidewalk segment, where we want to generate a random set of pedestrians but we want those pedestrians to remain the same between page refreshes or React component updates.

We can generate seeds by calling `generateRandSeed()`, then by feeding those seeds into the `seedrandom` library:

```js title="random-example.js"
import seedrandom from "seedrandom";
import { generateRandSeed } from "./util/random";

function doThingsWithThePRNG() {
  const seed = generateRandSeed();
  const prng1 = seedrandom(seed);

  // Generate pseudo-random numbers. Each time rng() is called, it returns
  // the next pseudo-random number in the sequence.
  console.log(prng1()); // -> first pseudo-random number
  console.log(prng1()); // -> second pseudo-random number
  // .. etc

  // Create another generator with the same seed
  const prng2 = seedrandom(seed);
  console.log(prng2()); // -> same value as the first call to prng1()
}
```

When `seedrandom()` is called, it returns a function that will return a consistent sequence of numbers each time it is called, for the given `seed`. The sequence returned is local to that function.

:::note

Although the `seed` can be any type of value, like a string, `generateRandSeed()` returns only integer values so that when we store seeds in a data model we can check for type consistency.

:::

In the above example, we call `seedrandom()` twice, with the same `seed`, so that we can create the same sequence of pseudo-random numbers in two different scenarios. This example is contrived because the pseudo-random number generators are used in the same function, and in reality the `seedrandom()` call may happen in separate parts of the application.

**References**

- [Random Seeds, Coded Hints, and Quintillions](http://davidbau.com/archives/2010/01/30/random_seeds_coded_hints_and_quintillions.html) [David Bau]
- [seedrandom.js](https://github.com/davidbau/seedrandom) [GitHub]

## axios

The [Axios](https://github.com/axios/axios) library was introduced as a recommendation over native `fetch` as a way to manage requests to the Streetmix back-end API layer. Additionally, Axios Mock Adapter offers the possibility to mock all API requests for integration testing.

If you work on API-related features, consider refactoring the `fetch` calls to use the API in `.assets/scripts/util/api.js`.
