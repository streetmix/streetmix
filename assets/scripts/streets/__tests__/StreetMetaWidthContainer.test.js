/* eslint-env jest */
import React from 'react'
import { screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { render } from '../../../../test/helpers/render'
import StreetMetaWidthContainer from '../StreetMetaWidthContainer'

describe('StreetMetaWidthContainer', () => {
  it('renders', () => {
    const { asFragment } = render(<StreetMetaWidthContainer />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders selection dropdown on click', async () => {
    const { asFragment } = render(<StreetMetaWidthContainer />)
    await userEvent.click(screen.getByText('0 m width'))
    expect(asFragment()).toMatchSnapshot()
  })

  it.skip('updates street label on selection change', async () => {
    const updateStreetWidth = jest.fn()
    const changeValue = 10

    // TODO: the `updateStreetWidth` function is not passed in because
    // `render()` does not allow props to override Redux connect
    // And the change value is called with the `40` not `changeValue`
    render(
      <StreetMetaWidthContainer
        street={{}}
        updateStreetWidth={updateStreetWidth}
      />
    )
    await userEvent.click(screen.getByTitle('Change width of the street'))
    userEvent.selectOptions(screen.getByRole('listbox'), changeValue)

    expect(updateStreetWidth).toBeCalledWith(changeValue)

    // Expect render to be label again
    expect(screen.getByRole('listbox')).toBe(null)
    expect(screen.getByTitle('Change width of the street')).not.toBe(null)
  })

  it.skip('does not render selection dropdown on click when not editable', async () => {
    // TODO: This should pass, but `editable={false}` is never set
    // Either update `render()` to allow props to override Redux connect,
    // or set initial state (even though it tightly couples component specification to Redux store)
    render(
      <StreetMetaWidthContainer
        street={{}}
        editable={false}
        updateStreetWidth={jest.fn()}
      />
    )
    await userEvent.click(screen.getByText('width', { exact: false }))

    expect(screen.getByRole('listbox')).toBe(null)
    expect(screen.getByText('width', { exact: false })).not.toBe(null)
  })

  // TODO: mock updateUnits() and test if it is called
  it.todo('updates units')
})
