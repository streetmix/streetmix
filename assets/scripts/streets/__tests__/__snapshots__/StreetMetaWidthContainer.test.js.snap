// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreetMetaWidthContainer renders 1`] = `
<DocumentFragment>
  <span
    class="street-metadata-width"
  >
    <span
      class="street-width street-width-editable"
    >
      <svg />
      0 m width
      <svg
        class="menu-carat-down"
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`StreetMetaWidthContainer renders selection dropdown on click 1`] = `
<DocumentFragment>
  <span
    class="street-metadata-width"
  >
    <select
      aria-describedby="tippy-3"
      class="street-width-select"
    >
      <option
        disabled=""
      >
        Occupied width:
      </option>
      <option
        disabled=""
      >
        0 m
      </option>
      <option
        disabled=""
      />
      <option
        disabled=""
      >
        Building-to-building width:
      </option>
      <option
        value="40"
      >
        12 m
      </option>
      <option
        value="60"
      >
        18 m
      </option>
      <option
        value="80"
      >
        24 m
      </option>
      <option
        disabled=""
      />
      <option>
        0 m
      </option>
      <option
        value="-1"
      >
        Different widthâ€¦
      </option>
      <option
        disabled=""
      />
      <option
        id="switch-to-imperial-units"
        value="-3"
      >
        Switch to imperial units (feet)
      </option>
      <option
        id="switch-to-metric-units"
        value="-2"
      >
        Switch to metric units
      </option>
    </select>
  </span>
</DocumentFragment>
`;
